sound/soc/codecs/es8326.c

```diff
@@ -0,0 +1,854 @@
+										return ret;
+						}
+				}
+				snd_soc_component_write(codec, ES8326_CLK_CTL_01, ES8326_CLK_ON);
+				snd_soc_component_write(codec, ES8326_RESET_00, ES8326_PWRUP_SEQ_EN);
+				snd_soc_component_write(codec, ES8326_INTOUT_IO_59, 0x45);
+				snd_soc_component_write(codec, ES8326_SDINOUT1_IO_5A,
+						(ES8326_IO_DMIC_CLK << ES8326_SDINOUT1_SHIFT));
+				snd_soc_component_write(codec, ES8326_SDINOUT23_IO_5B, ES8326_IO_INPUT);
+				snd_soc_component_write(codec, ES8326_CLK_RESAMPLE_03, 0x05);
+				snd_soc_component_write(codec, ES8326_HP_DRVIER_24, 0x00);
+				snd_soc_component_write(codec, ES8326_VMIDSEL_18, 0x02);
+				snd_soc_component_write(codec, ES8326_ANA_PDN_16, 0x00);
+				snd_soc_component_write(codec, ES8326_PGA_PDN_17, 0x40);
+				snd_soc_component_write(codec, ES8326_DAC2HPMIX_25, 0xAA);
+				snd_soc_component_write(codec, ES8326_ADC_MUTE_15, 0x00);
+				snd_soc_component_write(codec, ES8326_RESET_00, ES8326_CSM_ON);
+
+				break;
+		case SND_SOC_BIAS_STANDBY:
+				break;
+		case SND_SOC_BIAS_OFF:
+				if (es8326->mclk)
+						clk_disable_unprepare(es8326->mclk);
+				dev_dbg(codec->dev, "%s off\n", __func__);
+				snd_soc_component_write(codec, ES8326_ADC_MUTE_15, 0x1F);
+				snd_soc_component_write(codec, ES8326_DAC2HPMIX_25, 0x11);
+				snd_soc_component_write(codec, ES8326_RESET_00, ES8326_CSM_OFF);
+				snd_soc_component_write(codec, ES8326_PGA_PDN_17, 0xF8);
+				snd_soc_component_write(codec, ES8326_ANA_PDN_16, 0xFB);
+				snd_soc_component_write(codec, ES8326_VMIDSEL_18, 0x00);
+				snd_soc_component_write(codec, ES8326_HP_DRVIER_24, 0x0F);
+				snd_soc_component_write(codec, ES8326_INT_SOURCE_58, 0x08);
+				snd_soc_component_write(codec, ES8326_SDINOUT1_IO_5A, ES8326_IO_INPUT);
+				snd_soc_component_write(codec, ES8326_SDINOUT23_IO_5B, ES8326_IO_INPUT);
+				snd_soc_component_write(codec, ES8326_RESET_00, ES8326_CODEC_RESET|ES8326_PWRUP_SEQ_EN);
+				snd_soc_component_write(codec, ES8326_CLK_CTL_01, ES8326_CLK_OFF);
+				break;
+		}
+
+		return 0;
+}
+
+#define es8326_RATES SNDRV_PCM_RATE_8000_96000
+
+#define es8326_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S20_3LE |\
+	SNDRV_PCM_FMTBIT_S24_LE)
+
+static struct snd_soc_dai_ops es8326_ops = {
+		.hw_params = es8326_pcm_hw_params,
+		.set_fmt = es8326_set_dai_fmt,
+		.set_sysclk = es8326_set_dai_sysclk,
+};
+
+static struct snd_soc_dai_driver es8326_dai = {
+		.name = "ES8326 HiFi",
+		.playback = {
+				.stream_name = "Playback",
+				.channels_min = 1,
+				.channels_max = 2,
+				.rates = es8326_RATES,
+				.formats = es8326_FORMATS,
+				},
+		.capture = {
+				.stream_name = "Capture",
+				.channels_min = 1,
+				.channels_max = 2,
+				.rates = es8326_RATES,
+				.formats = es8326_FORMATS,
+				},
+		.ops = &es8326_ops,
+		.symmetric_rates = 1,
+};
+
+static int es8326_probe(struct snd_soc_component*codec)
+{
+		struct es8326_priv *es8326 = snd_soc_component_get_drvdata(codec);
+		es8326->component = codec;
+
+		if (codec == NULL) {
```

Comment:
Can this happen?

Also even if it's needed, you have an NULL dereference on line 606